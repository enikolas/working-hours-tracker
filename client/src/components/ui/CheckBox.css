.check-box label {
  font-size: 10px;
  line-height: 11px;
  display: block;
}

.check-box:before {
  content: ' ';
  display: block;
  font-size: 16px;
  line-height: 18px;
  text-align: center;
  background: #fff;
  border-radius: 5px;
  width: 16px;
  height: 16px;
  border: 2px solid #c5ccd1;
  float: left;
  margin-right: 8px;
  color: #254356;
}

.check-box.checked:before {
  content: 'âœ“';
}

.check-box:focus {
  outline: none;
}

.check-box.not(.disabled):focus:before {
  border: 2px solid #f93;
}

.check-box.disabled {
  opacity: 0.5;
}

.no-pause {
  width: 100px;
  position: absolute;
  margin-left: 320px;
}
