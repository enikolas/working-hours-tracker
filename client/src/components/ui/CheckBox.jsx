import React from 'react';
import PropTypes from 'prop-types';

import './CheckBox.css';

const KEY_SPACEBAR = 32;

class CheckBox extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      checked: props.value
    };
    this.handleCheck = this.handleCheck.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.value !== this.props.value) {
      this.setState({ checked: nextProps.value });
    }
  }

  handleCheck(e) {
    // if it is not click, then just spacebar toggles it
    if (this.props.disabled || (e.type !== 'click' && e.keyCode !== KEY_SPACEBAR)) {
      return;
    }
    e.preventDefault();
    this.setState((state) => {
      const newChecked = !state.checked;
      this.props.onCheck(newChecked);
      return { checked: newChecked };
    });
  }

  render() {
    const { checked } = this.state;
    const { label, disabled } = this.props;
    const className = `check-box${checked ? ' checked' : ''}${disabled ? ' disabled' : ''}`;
    return (
      <div
        role="checkbox"
        className={className}
        onClick={this.handleCheck}
        onKeyDown={this.handleCheck}
        aria-checked={checked}
        tabIndex={0}
      >
        <label>{label}</label>
      </div>
    );
  }
}

CheckBox.propTypes = {
  value: PropTypes.bool,
  disabled: PropTypes.bool,
  onCheck: PropTypes.func,
  label: PropTypes.string.isRequired
};

CheckBox.defaultProps = {
  value: false,
  disabled: false,
  onCheck: () => {}
};

export default CheckBox;
