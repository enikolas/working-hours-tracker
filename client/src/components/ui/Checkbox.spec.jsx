import React from 'react';
import { shallow } from 'enzyme';

import CheckBox from './CheckBox';

const KEY_SPACEBAR = 32;

describe('CheckBox', () => {
	describe('render', () => {
		it('should render properly when minimum props are passed', () => {
			const wrapper = shallow(<CheckBox label="hello check" />);
			expect(wrapper).toMatchSnapshot();

			const wrapperDisabled = shallow(<CheckBox label="hello check" disabled />);
			expect(wrapperDisabled).toMatchSnapshot();

			const wrapperChecked = shallow(<CheckBox label="hello check" value />);
			expect(wrapperChecked).toMatchSnapshot();
		});
	});
	describe('handleCheck', () => {
		it('should mark as checked when spacebar is pressed', () => {
			const wrapper = shallow(<CheckBox label="hello check" />);
			wrapper.simulate('keyDown', {
				preventDefault: jest.fn(),
				type: 'keyDown',
				keyCode: KEY_SPACEBAR
			});
			expect(wrapper.state('checked')).toEqual(true);
		});
		it('should not mark if props is disabled', () => {
			const preventDefault = jest.fn();
			const wrapper = shallow(<CheckBox label="hello check" disabled />);
			wrapper.simulate('keyDown', {
				preventDefault,
				type: 'keyDown',
				keyCode: KEY_SPACEBAR
			});
			expect(preventDefault).not.toHaveBeenCalled();
		});
		it('should not mark if key press is not space', () => {
			const preventDefault = jest.fn();
			const wrapper = shallow(<CheckBox label="hello check" />);
			wrapper.simulate('keyDown', {
				preventDefault,
				type: 'keyDown',
				keyCode: 0
			});
			expect(preventDefault).not.toHaveBeenCalled();
		});
	});
	describe('componentWillReceiveProps', () => {
		it('should change its state if props value change', () => {
			const wrapper = shallow(<CheckBox label="hello check" />);
			expect(wrapper.state('checked')).toBeFalsy();
			wrapper.setProps({ value: true });
			expect(wrapper.state('checked')).toBeTruthy();
		});
	});
});
