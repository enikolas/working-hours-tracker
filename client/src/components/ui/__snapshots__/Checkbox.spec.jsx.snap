// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBox render should render properly when minimum props are passed 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CheckBox
    disabled={false}
    label="hello check"
    onCheck={[Function]}
    value={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "aria-checked": false,
      "children": <label>
        hello check
      </label>,
      "className": "check-box",
      "onClick": [Function],
      "onKeyDown": [Function],
      "role": "checkbox",
      "tabIndex": 0,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "hello check",
      },
      "ref": null,
      "rendered": "hello check",
      "type": "label",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "aria-checked": false,
        "children": <label>
          hello check
        </label>,
        "className": "check-box",
        "onClick": [Function],
        "onKeyDown": [Function],
        "role": "checkbox",
        "tabIndex": 0,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "hello check",
        },
        "ref": null,
        "rendered": "hello check",
        "type": "label",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`CheckBox render should render properly when minimum props are passed 2`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CheckBox
    disabled={true}
    label="hello check"
    onCheck={[Function]}
    value={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "aria-checked": false,
      "children": <label>
        hello check
      </label>,
      "className": "check-box disabled",
      "onClick": [Function],
      "onKeyDown": [Function],
      "role": "checkbox",
      "tabIndex": 0,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "hello check",
      },
      "ref": null,
      "rendered": "hello check",
      "type": "label",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "aria-checked": false,
        "children": <label>
          hello check
        </label>,
        "className": "check-box disabled",
        "onClick": [Function],
        "onKeyDown": [Function],
        "role": "checkbox",
        "tabIndex": 0,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "hello check",
        },
        "ref": null,
        "rendered": "hello check",
        "type": "label",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`CheckBox render should render properly when minimum props are passed 3`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CheckBox
    disabled={false}
    label="hello check"
    onCheck={[Function]}
    value={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "aria-checked": true,
      "children": <label>
        hello check
      </label>,
      "className": "check-box checked",
      "onClick": [Function],
      "onKeyDown": [Function],
      "role": "checkbox",
      "tabIndex": 0,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "hello check",
      },
      "ref": null,
      "rendered": "hello check",
      "type": "label",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "aria-checked": true,
        "children": <label>
          hello check
        </label>,
        "className": "check-box checked",
        "onClick": [Function],
        "onKeyDown": [Function],
        "role": "checkbox",
        "tabIndex": 0,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "hello check",
        },
        "ref": null,
        "rendered": "hello check",
        "type": "label",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
